desc/objectives/_free_boundary.py

```diff
@@ -0,0 +1,882 @@
+            fi = jnp.abs(fi) if abserr else fi
+            wi = w[i * nn : (i + 1) * nn]
+            fmax = jnp.max(fi)
+            fmin = jnp.min(fi)
+            fmean = jnp.mean(fi * wi) / jnp.mean(wi)
+            _print(fmt, fmax, fmin, fmean, unit)
+
+
+class BoundaryError(_Objective):
+    """Target for free boundary conditions on LCFS for finite beta equilibrium.
+
+    Computes the residual of the following:
+
+    𝐁ₒᵤₜ ⋅ 𝐧 = 0
+    𝐁ₒᵤₜ² - 𝐁ᵢₙ² - p = 0
+    μ₀∇Φ − 𝐧 × [𝐁ₒᵤₜ − 𝐁ᵢₙ]
+
+    Where 𝐁ᵢₙ is the total field inside the LCFS (from fixed boundary calculation)
+    𝐁ₒᵤₜ is the total field outside the LCFS (from coils and virtual casing principle),
+    𝐧 is the outward surface normal, p is the plasma pressure, and Φ is the surface
+    current potential on the LCFS.
+
+    The third equation is only included if a sheet current is supplied, otherwise it
+    is trivially satisfied. If it is known that the external field accurately reproduces
+    the target equilibrium with low normal field error and pressure at the edge is zero,
+    then the sheet current will generally be negligible and can be omitted to save
+    effort.
+
+    This objective also works for vacuum equilibria, though in that case
+    VacuumBoundaryError will be much faster as it avoids the singular virtual casing
+    integral.
+
+    Parameters
+    ----------
+    eq : Equilibrium
+        Equilibrium that will be optimized to satisfy the Objective.
+    ext_field : MagneticField
+        External field produced by coils.
+    sheet_current : FourierCurrentPotentialField
+        Sheet current on the last closed flux surface. Not required for vacuum fields,
+        but generally needed to correctly solve at finite beta/current. Geometry will
+        be automatically constrained to be the same as the equilibrium LCFS.
+    target : float, ndarray, optional
+        Target value(s) of the objective. Only used if bounds is None.
+        len(target) must be equal to Objective.dim_f
+    bounds : tuple, optional
+        Lower and upper bounds on the objective. Overrides target.
+        len(bounds[0]) and len(bounds[1]) must be equal to Objective.dim_f
+    weight : float, ndarray, optional
+        Weighting to apply to the Objective, relative to other Objectives.
+        len(weight) must be equal to Objective.dim_f
+    normalize : bool
+        Whether to compute the error in physical units or non-dimensionalize.
+    normalize_target : bool
+        Whether target and bounds should be normalized before comparing to computed
+        values. If `normalize` is `True` and the target is in physical units,
+        this should also be set to True.
+    s, q : integer
+        Hyperparameters for singular integration scheme, s is roughly equal to the size
+        of the local singular grid with respect to the global grid, q is the order of
+        integration on the local grid
+    src_grid, eval_grid : Grid, optional
```

Comment:
may be worth specifying in the doc that the `src_grid` must be `sym=False`, and that the `src_grid` is specifically for evaluating the singular integral on the plasma surface